### PR TITLE
Benchmark GitHub Actions workflow

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

Benchmark GitHub Actions workflow.

[Workflow run page](https://github.com/huggingface/transformers/actions/runs/9318192661/job/25650027316)
[Dataset on the Hub](https://huggingface.co/datasets/hf-internal-testing/benchmark_results/tree/main)

(Not done yet)
- Maybe change decoding steps from 128 to 1024
- Run against Llama2 too

We are mostly only interested in `summary.json` file.

However, I am uploading the whole directory of an experiment in each run. The reason is to keep the benchmark config files available if we ever need to access this information. However, this seems a bit too much (too many files for which most of time they remain the same across different dates)

### CODE REVIEW

File: .github/workflows/benchmark.yml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+      - name: Reinstall transformers in edit mode (remove the one installed during docker image build)
+        working-directory: /transformers
+        run: python3 -m pip uninstall -y transformers && python3 -m pip install -e .
+
+      - name: Benchmark
+        working-directory: /transformers
+        run: |
+          python3 -m pip install optimum-benchmark>=0.2.0
+          HF_TOKEN=${{ secrets.TRANSFORMERS_HUB_BOT_HF_TOKEN }} python3 benchmark/benchmark.py --repo_id hf-internal-testing/benchmark_results --path_in_repo $(date +'%Y-%m-%d') --config-dir benchmark/config --config-name generation --commit=${{ github.sha }} backend.model=google/gemma-2b backend.cache_implementation=null,static backend.torch_compile=false,true --multirun
```
Comment (ArthurZucker) [inline .github/workflows/benchmark.yml:33]: can we post it somewhere / run a python script to compare the results from previous commit / from average of previous commits?

Comment (ArthurZucker) [inline .github/workflows/benchmark.yml:33]: let's be super careful and have a finegrained token for that!

**Review by ArthurZucker — COMMENTED** [review]
Looks great already!

File: .github/workflows/benchmark.yml (hunk 1 above)
Comment (ydshieh) [reply .github/workflows/benchmark.yml:33]: I am considering doing this with a Space app which will fetch the results from this dataset and show some graph.
But if you think we should also do such comparison within the same workflow run, I can add something.
(so far the dataset is kinda empty, so maybe better I add that part one day in the next week?)

Comment (ydshieh) [reply .github/workflows/benchmark.yml]: Update to a new secret `TRANSFORMERS_BENCHMARK_TOKEN` (finegrained token)

Comment (ArthurZucker) [reply .github/workflows/benchmark.yml]: ok sounds good

**Review by ArthurZucker — COMMENTED** [review]
Almost good to go, let's run it when important model slow is triggered as well

File: .github/workflows/benchmark.yml [hunk 2]
```diff
@@ -0,0 +1,33 @@
+name: Self-hosted runner (benchmark)
+
+on:
+  schedule:
+    - cron: "17 2 * * *"
```
[the remark this answers is not in the document]
Comment (ydshieh) [reply .github/workflows/benchmark.yml:5]: Yeah, I agree. I would instead push the results to another dataset (one for daily CI, one for push to main event)

[the remark this answers is not in the document]
Comment (ydshieh) [reply .github/workflows/benchmark.yml:5]: update to add the following new block:

- for `.github/workflows/push-important-models.yml` being triggered
- upload to `hf-internal-testing/benchmark_results_merge_event`


```
      - name: Benchmark (merged to main event)
        if: github.event_name == 'push' && github.ref_name == 'main'
        working-directory: /transformers
        run: |
          python3 -m pip install optimum-benchmark>=0.2.0
          HF_TOKEN=${{ secrets.TRANSFORMERS_BENCHMARK_TOKEN }} python3 benchmark/benchmark.py --repo_id hf-internal-testing/benchmark_results_merge_event --path_in_repo $(date +'%Y-%m-%d') --config-dir benchmark/config --config-name generation --commit=${{ github.sha }} backend.model=google/gemma-2b backend.cache_implementation=null,static backend.torch_compile=false,true --multirun
```

**Review by ArthurZucker — APPROVED** [review]
Thanks! 🚀

**Pull request merged**